### PR TITLE
Add maxConcurrency flag to optimize multiple images with limited concurrency

### LINKED ISSUE #30: Runs out of memory when optimizing a large number of files
## The problem

When attempting to optimize XXX png files, we received the following error:

> Error at makeError (/work/ios-sdk/node_modules/imagemin-pngquant/node_modules/execa/lib/error.js:59:11) at handlePromise (/work/ios-sdk/node_modules/imagemin-pngquant/node_modules/execa/index.js:114:26) at runMicrotasks (<anonymous>) at processTicksAndRejections (internal/process/task_queues.js:97:5) at async /work/ios-sdk/node_modules/p-pipe/index.js:12:19 at async handleFile (/work/ios-sdk/node_modules/imagemin/index.js:21:9) at async /work/ios-sdk/node_modules/imagemin/index.js:54:13 at async Promise.all (index 11) at async Promise.all (index 0) at async module.exports (/work/ios-sdk/node_modules/@mapbox/appropriate-images/node_modules/p-finally/index.js:9:11)npm ERR! code ELIFECYCLEnpm ERR! errno 1npm ERR! ios-sdk@0.0.0 build-images: `mkdir -p src/img/dist && node bin/build-image-config.js && node bin/appropriate-images.js --all`npm ERR! Exit status 1npm ERR!npm ERR! Failed at the ios-sdk@0.0.0 build-images script.npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

After some prodding, we were able to return a more useful error message: imagemin is running out of memory:

> Error: Command failed with ENOMEM: /work/ios-sdk/node_modules/pngquant-bin/vendor/pngquant -
> spawn ENOMEM
>  at ChildProcess.spawn (internal/child_process.js:407:11)
>  at Object.spawn (child_process.js:548:9)
>  at execa (/work/ios-sdk/node_modules/imagemin-pngquant/node_modules/execa/index.js:80:26)
>  at /work/ios-sdk/node_modules/imagemin-pngquant/index.js:57:21
>  at /work/ios-sdk/node_modules/p-pipe/index.js:12:25
>  at handleFile (/work/ios-sdk/node_modules/imagemin/index.js:21:54)
>  at async /work/ios-sdk/node_modules/imagemin/index.js:54:13
>  at async Promise.all (index 209)
>  at async module.exports (/work/ios-sdk/node_modules/@mapbox/appropriate-images/lib/optimize.js:47:34)
>  at async module.exports (/work/ios-sdk/node_modules/@mapbox/appropriate-images/node_modules/p-finally/index.js:9:11)

## Proposed solution

In https://stackoverflow.com/a/63573743, the poster found success by using p-limit:

> The solution for us, was to use plimit to limit the number of concurrent promises to 5 or 10. On the same number of images (20k+) the VIRT mem usage never rose, neither did RES, and we stopped getting ENOMEM errors.

I tested this by optimizing each file and and using p-limit and it worked.

Example:

```js
    const makeItHappen = createTemporaryDirectory()
      .then(() => Promise.all(generateSizeVariants))
      .then((result) => {
        if (configErrors.length) return Promise.reject(configErrors);
        return result;
      })
      .then((result) => {
        const optimizeQueue = [];

        for (var res in result) {
          optimizeQueue.push(
            limit(() => optimize(_.flatten(result[res]), options))
          );
        }
        return Promise.all(optimizeQueue);
      });
```

## Next steps

We should consider adding a max concurrency flag that, when enabled, will use p-limit to handle optimizing each file.

cc @colleenmcginnis

### PR DESCRIPTION
**Pull request by katydecorah**
This PR adds a new flag, `maxConcurrency` or `c`. If set, then each file will be optimized using the given number of limited concurrency (instead of all files being optimized at once). This will help prevent imagemin from running out of memory when processing a large volume of images.

This PR also updates a few dependencies for patch fixes.

Fixes #30

### CODE REVIEW

**Comment by katydecorah** [comment]
@colleenmcginnis would you be up for reviewing? Here are some suggestions on how to test:

1. Pull down https://github.com/mapbox/ios-sdk/pull/1065 (since this repo is public, I installed appropriate-images from this branch.)
    - [ ] Push https://github.com/mapbox/ios-sdk/pull/1065 to staging to confirm that the build succeeds. (I pushed the branch and it succeeded, but I also just updated the branch from production, so you could push the latest commit to staging to double confirm that the build will succeed.)
2. Try testing the `build-images` script:
    - [ ] Run `npm run build-images` to make sure the script generates images with the concurrency flag.
    - [ ] Remove `-c 10` in package.json build-images script and then run `npm run build-images` again to make sure the script generates images as expected with out the concurrency flag.

File: lib/create-cli.js [hunk 1]
```diff
@@ -76,14 +80,18 @@ module.exports = (imageConfig, options) => {
...
+
   const logUsageError = (error) => {
     console.log(`${chalk.red.bold('Usage error:')} ${error.message}`);
   };
 
   generate(imageConfig, generateOptions)
     .then((filenames) => {
       if (cli.flags.quiet) return;
-      const printableFilenames = filenames
+      const printableFilenames = _.flatten(filenames)
```
Comment (colleenmcginnis) [inline lib/create-cli.js:94]: What is this change doing?

File: lib/generate.js [hunk 2]
```diff
@@ -80,6 +81,26 @@ function clearPriorOutput(directory, sourceImageBasename) {
...
+  if (options.maxConcurrency) {
+    const limit = pLimit(options.maxConcurrency);
+    const optimizeQueue = [];
+    for (const res in results) {
+      optimizeQueue.push(
+        limit(() => optimize(_.flatten(results[res]), options))
+      );
+    }
+    return Promise.all(optimizeQueue);
+  } else return optimize(_.flatten(results), options);
```
Comment (colleenmcginnis) [inline lib/generate.js:101]: Do you need `{}` around the return? I've always done it that way if the `if` includes `{}`, but maybe it's not necessary. Like this?

```suggestion
  } else {
    return optimize(_.flatten(results), options);
  }
```

File: test/generate.test.js [hunk 3]
```diff
@@ -69,6 +69,52 @@ describe('generate', () => {
...
+        basename: 'osprey.jpg',
+        sizes: [{ width: 300 }, { width: 600 }]
+      },
+      walrus: {
+        basename: 'walrus.png',
+        sizes: [{ width: 300 }, { width: 600 }]
+      }
+    };
+
+    const options = { inputDirectory, outputDirectory, maxConcurrency: 5 };
```
Comment (colleenmcginnis) [inline test/generate.test.js:92]: Should we test using a `maxConcurrency` that's lower than the total number of images being processed or does it not matter?

**Review by colleenmcginnis — APPROVED** [review]
@katydecorah I tested the ios-sdk branch as you suggested and everything worked as expected. I left a couple minor questions where I wasn't sure why something was the way it was.

File: lib/generate.js (hunk 2 above)
Comment (katydecorah) [reply lib/generate.js:101]: @colleenmcginnis It isn't necessary if the value is all on one line, but I can add the curly brackets to improve legibility!

File: lib/create-cli.js (hunk 1 above)
Comment (katydecorah) [reply lib/create-cli.js:94]: @colleenmcginnis This is flattening the array of filenames, which is needed when processing multiple files separately.

File: test/generate.test.js (hunk 3 above)
Comment (katydecorah) [reply test/generate.test.js:92]: @colleenmcginnis Yes, I think that would make for a better test cases. Updated!

**Pull request merged**